### DISCUSSION TITLE
Failure to find file when building docs

### DISCUSSION THREAD

**Discussion by snayff** [community/community#26873] [Actions]
<p>I am trying to build Sphinx documentation for my python project using Github Actions and I am hitting an error stating that the relevant file is not found. The documentation builds fine locally (it works on my machine!). I thought it was might be an issue with location of file but that doesnt seem to be the case as from as the paths in the error hold the correct relative position.</p>
<p>I’ve checked the Sphinx and Github Actions documentation and I havent found anything to indicate what I am doing wrong.</p>
<p>The file structure looks like this:</p>
<ul>
<li>data
<ul>
<li>game
<ul>
<li>traits.json  &lt;- the file not found</li>
</ul>
</li>
</ul>
</li>
<li>docs
<ul>
<li>_build &lt;- where the docs are being built to</li>
</ul>
</li>
<li>scripts
<ul>
<li>engine
<ul>
<li>action.py  &lt;- where the error occurs</li>
</ul>
</li>
</ul>
</li>
</ul>
<p>The error:</p>
<pre><code class="lang-python">Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.8.5/x64/lib/python3.8/site-packages/sphinx/ext/autodoc/importer.py", line 66, in import_module
    return importlib.import_module(modname)
  File "/opt/hostedtoolcache/Python/3.8.5/x64/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "&lt;frozen importlib._bootstrap&gt;", line 1014, in _gcd_import
  File "&lt;frozen importlib._bootstrap&gt;", line 991, in _find_and_load
  File "&lt;frozen importlib._bootstrap&gt;", line 975, in _find_and_load_unlocked
  File "&lt;frozen importlib._bootstrap&gt;", line 671, in _load_unlocked
  File "&lt;frozen importlib._bootstrap_external&gt;", line 783, in exec_module
  File "&lt;frozen importlib._bootstrap&gt;", line 219, in _call_with_frames_removed
  File "/home/runner/work/notquiteparadise/notquiteparadise/scripts/engine/action.py", line 26, in &lt;module&gt;
    from scripts.engine.effect import Effect, MoveActorEffect
  File "/home/runner/work/notquiteparadise/notquiteparadise/scripts/engine/effect.py", line 9, in &lt;module&gt;
    from scripts.engine import utility, world
  File "/home/runner/work/notquiteparadise/notquiteparadise/scripts/engine/world.py", line 15, in &lt;module&gt;
    from scripts.engine import action, chronicle, debug, library, utility
  File "/home/runner/work/notquiteparadise/notquiteparadise/scripts/engine/chronicle.py", line 8, in &lt;module&gt;
    from scripts.engine import library, world
  File "/home/runner/work/notquiteparadise/notquiteparadise/scripts/engine/library.py", line 228, in &lt;module&gt;
    refresh_library()
  File "/home/runner/work/notquiteparadise/notquiteparadise/scripts/engine/library.py", line 73, in refresh_library
    _load_traits_data()
  File "/home/runner/work/notquiteparadise/notquiteparadise/scripts/engine/library.py", line 111, in _load_traits_data
    with open(DATA_PATH + "game/traits.json") as file:
FileNotFoundError: [Errno 2] No such file or directory: '/home/runner/work/notquiteparadise/notquiteparadise/data/game/traits.json'
</code></pre>
<p>The Github Action yaml:</p>
<pre><code class="lang-auto">name: NQP - Build Docs

on:
  workflow_run:
    workflows: ["NQP - Run Tests"]
    branches: [develop, master]
    types:
      - completed

jobs:
  docs:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v2

      - name: Set up Python 3.8
        uses: actions/setup-python@v2
        with:
          python-version: 3.8

      - name: Use cache
        uses: actions/cache@v2
        with:
          path: ~\AppData\Local\pip\Cache
          key: ${{ runner.os }}-pip-${{ hashFiles('**/poetry.lock') }}
          restore-keys: ${{ runner.os }}-pip-

      - name: Install dependencies
        run: |
          sudo apt-get update
          sudo apt-get install libsdl2-dev
          sudo apt-get install python-sphinx
          python -m pip install --upgrade poetry
          poetry config virtualenvs.create false
          poetry install --no-root

      - name: build-docs
        run: |
          sphinx-apidoc -o ./docs .
          cd docs
          sphinx-build -a -E --keep-going . _build
          
      - name: deploy-docs
        uses: JamesIves/github-pages-deploy-action@3.5.7
        with:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          BRANCH: gh-pages
          FOLDER: docs/_build

</code></pre>
<p>Any help would be appeciated!</p>

**Comment by ghostwriter** [comment, marked as answer]
<p>Your code was referencing a directory that does not exist.</p>
<p>sent you a PR to correct this.</p>
<p>Fix:</p>
<pre><code class="lang-auto">git mv data/Game data/temp

git mv data/temp data/game

git commit -m "Rename folder Game to game"
</code></pre>
<p>Hope that’s helpful :v:t4:</p>

**Comment by snayff** [comment]
<p>Thank you, mate, I have no idea why it was ever Game instead of game!</p>
<p>The PR seems to have fixed that issue (thank you!) but generated a new one. I’ve not seen anything like that before, have you?</p>
<aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/Snayff/notquiteparadise/runs/1143276508?check_suite_focus=true#step:7:38" target="_blank" rel="noopener nofollow ugc">GitHub</a>
  </header>
  <article class="onebox-body">
    <img src="https://user-images.githubusercontent.com/23128268/181100407-bef5e83e-e6a8-48c7-a04d-1b7466b45ceb.jpeg" class="thumbnail onebox-avatar" width="400" height="400">

<h3><a href="https://github.com/Snayff/notquiteparadise/runs/1143276508?check_suite_focus=true#step:7:38" target="_blank" rel="noopener nofollow ugc">Snayff/notquiteparadise</a></h3>

<p>Possibly, one day, a roguelike. Contribute to Snayff/notquiteparadise development by creating an account on GitHub.</p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<pre><code class="lang-auto">ALSA lib confmisc.c:767:(parse_card) cannot find card '0'
ALSA lib conf.c:4528:(_snd_config_evaluate) function snd_func_card_driver returned error: No such file or directory
ALSA lib confmisc.c:392:(snd_func_concat) error evaluating strings
ALSA lib conf.c:4528:(_snd_config_evaluate) function snd_func_concat returned error: No such file or directory
ALSA lib confmisc.c:1246:(snd_func_refer) error evaluating name
ALSA lib conf.c:4528:(_snd_config_evaluate) function snd_func_refer returned error: No such file or directory
ALSA lib conf.c:5007:(snd_config_expand) Evaluate error: No such file or directory
ALSA lib pcm.c:2495:(snd_pcm_open_noupdate) Unknown PCM default
</code></pre>
<pre><code class="lang-auto">File "/home/runner/work/notquiteparadise/notquiteparadise/scripts/engine/ui/manager.py", line 65, in __init__
    self._window: pygame.display = pygame.display.set_mode((self._desired_width, self._desired_height))
pygame.error: No available video device
</code></pre>

**Comment by airtower-luna** [comment]
<p>If that’s from the Actions build, too, I’d guess it’s tests having issues with the runner VMs not having displays and sound cards. Could you run tests in a headless mode? If not you might need to set up some sort of emulation.</p>

**Comment by snayff** [comment]
<p>I think all the tests are headless but given this is just to build documentation I would have thought it was all headless anyway.</p>
<p>I checked the pygame repo for their CI build and it flags video drivers so perhaps I need to do that, too. They’re using Travis so I need to check if there is anything that wouldn’t apply to Actions. <a href="https://github.com/pygame/pygame/blob/d43040e3825535bb543d72a270f1fa3a591875aa/.travis.yml" rel="noopener nofollow ugc">https://github.com/pygame/pygame/blob/d43040e3825535bb543d72a270f1fa3a591875aa/.travis.yml</a></p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="Snayff" data-post="6" data-topic="133210">
<div class="title">
<div class="quote-controls"></div>
 Snayff:</div>
<blockquote>
<p>I think all the tests are headless but given this is just to build documentation I would have thought it was all headless anyway.</p>
</blockquote>
</aside>
<p>That depends: As <a href="https://www.sphinx-doc.org/en/master/man/sphinx-apidoc.html">noted in the documentation</a> the call to <code>sphinx-apidoc</code> will import the modules, and side-effects of the imports will be executed. So if, for example, any of your modules tries to initialize display access during import that won’t work headless, similar for audio. The <code>pygame.display.set_mode()</code> call hints at something like that.</p>
<p>If that is indeed the issue the ideal fix would be to change the module so that display and audio initialization is done only when any related code is called.</p>

**Comment by snayff** [comment]
<p>Ah, I am with you. That makes sense and it gives me a clear way forward. Thank you very much. I’ll refactor to avoid those sort of calls and see how it goes.</p>

**Comment by snayff** [comment]
<p>Legend, that did it! Thank you again!</p>

**Comment by airtower-luna** [comment]
<p>Nice, thanks for the update! :smile_cat:</p>
